doc/design/reader/low_level_reader.md

```diff
@@ -0,0 +1,71 @@
+# Paddle Low Level Data Reader
+
+This design doc will describe a way to implement C++ written data reader(feeder) to
+feed data to paddle when train or inference inorder to improve performance, and make
+best usage of CPU and GPU usage.
+
+This design can be used for both V2 API and new operator design.
+
+## Background
+
+We have sevaral issues talking about training with v2 API get bad training performance,
+like:
+- https://github.com/PaddlePaddle/Paddle/issues/3675 and
+- https://github.com/PaddlePaddle/Paddle/issues/4156 and
+- https://github.com/PaddlePaddle/Paddle/issues/3600.
+
+The current implementations for data reader can greatly simplify user defining how
+to parse and feed the data to neural network, but we need to copy data from python
+to C++ when feeding the data, and the data feeder will wait until the calculations
+of one mini-batch is done.
+
+In V1 implementations there are some considerations to improve data feeding
+performance, we can port these features to improve the reader.
+
+## Implementaion
+
+A low-level reader contains below components:
+
+1. Data parser:
+    A data parser will use `dlopen` to open user-defined parser plugin, and call
+    `parse` to convert a line of string to internal `Matrix` or `Tensor` object.
+1. User parser plugin:
+    User must write a C++ plugin to parse data to `Matrix` or `Tensor` object. The
+    plugin should implement one interface:
+    ```c++
+    // return Matrix* for v2 API call
+    Tensor* parse(const std::string& line);
```

Comment:
Thanks for such detailed commonts.

> Here does user return one sample or a sample batch?
> I guess here user returns a single sample? Am I right?

You are right. User only need to define a "parser" to parse ***one*** raw sample data to a `Tensor`.

> Who will do the shuffle?

Let paddle do it of course. Low-level reader is not that "low-level" to let user do everything. The user interface should be the same as python, except use C++.

> Are the buffer and thread pool all implemented by the users or by Paddle?

Same as above, let paddle implement it.

---

> Does this mean we have to expose too many details about how data organized in Paddle's C++ codes? Is this necessary?

We only expose `Tensor` and `Vector`(for v1/v2)

> I am not very sure, from this interface Tensor* parse(const std::string& line); it seems that the user directly returns a data batch, does this means the user should allocate the memory and release it correctly?

User should allocate memory, but should not release it, let paddle do the job.

>If a data batch is directly returned, does this mean the user should also consider shuffling the data, and returned an already shuffled data batch? Shuffle is now in Python reader decorator. Do we intend to leave all these details to the user?

Buffering, queueing, shuffle should be done by paddle, like ` reader = api.low_level_reader(parser="myparser.so", buffer_size=8192, buffer_type="mmap", batch_size=128, shuffle=True)`

Reply:
I see, thank you~

Reply:
I like the interface:
```
reader = api.low_level_reader(parser="myparser.so", buffer_size=8192, buffer_type="mmap", batch_size=128, shuffle=True)
```
A little uncertain about if we need to allow user providing custom *.so file, because it has its associated problems:

1. User need to compile a so file for the same architecture.
1. Paddle Cloud after refactor is just a runtime engine that execute a user-define graph, no user code is running on it. But allowing custom so file will allow user code to run on Paddle cloud after refactor.

I agree that *.so file is more flexible, however, I am curious if we implement some common parsing OP, will it be sufficient? Like:
```
reader = api.low_level_reader(parser=paddle.substract_mean, buffer_size=8192, batch_size=128, shuffle=True)
```
How do you think? @lcy-seso @typhoonzero

Reply:
I am also concerned to compile a `*.so` is not very convenient. Personally, I prefer to implement some common parsing OP, and enhance the buffering, queueing, shuffle in the PaddlePaddle end which can be transparent to users. The low-level API can be exposed to users if they are extremely concerns about the accelerating the I-O speed.

Reply:
Common parsers are fine, like parsers to process specific datasets. We can gather these parsers implements gradually, and replace data readers under `python/paddle/v2/datasets`. 

For normal jobs which reader's performance does not harm, they can use: https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/reader/decorator.py#L162

Some cases like job can not get >80% of GPU performance can consider using low-level reader.